Support custom allocators in `Box`

TimDiekmann opened the pull request:
r? @Amanieu 

This pull request requires a crater run.

### Prior work:
- #71873
- #58457
- [`alloc-wg`](https://github.com/TimDiekmann/alloc-wg)-crate

Currently blocked on:
- ~#77118~
- ~https://github.com/rust-lang/chalk/issues/615 (#77515)~

TimDiekmann commented:
@nikomatsakis wrote in https://github.com/rust-lang/rust/pull/71873#issuecomment-632116962
> The chalk error is legit in the sense that we haven't defined what fundamental means for a type with multiple type parameters. I guess we will ignore all but the first, but we're going to have to spend at least some time thinking about how to use this to break coherence :)

What is the state of this?

jackh726 commented:
> @nikomatsakis wrote in [#71873 (comment)](https://github.com/rust-lang/rust/pull/71873#issuecomment-632116962)
> 
> > The chalk error is legit in the sense that we haven't defined what fundamental means for a type with multiple type parameters. I guess we will ignore all but the first, but we're going to have to spend at least some time thinking about how to use this to break coherence :)
> 
> What is the state of this?

No change. But I'll be sure to bring this up in the wg-traits meeting on tuesday.

TimDiekmann commented:
That would be great, thank you!

TimDiekmann commented:
Rebased onto master to include #77118

nagisa commented on library/alloc/src/boxed.rs:
```diff
@@ -518,20 +751,47 @@ impl<T: ?Sized> Box<T> {
+        A: 'a,
+    {
+        let manually_drop = mem::ManuallyDrop::new(b);
+        let alloc = unsafe { ptr::read(&manually_drop.1) };
```
You probably want `MaybeUninit` as you're effectively un-initializing part of the data contained within `ManuallyDrop` which is somewhat sketchy as the value is later used again to obtain `reference`.

TimDiekmann replied:
Ah, nice catch. So I think this way this should be safe?
```rust
let mut boxed = mem::MaybeUninit::new(b);
unsafe {
    let reference = &mut *(*b.as_mut_ptr()).0.as_ptr();
    let alloc = ptr::read(&(*b.as_mut_ptr()).1);
    (reference, alloc)
}
```

Or would it also be safe to obtain the reference **before** reading the allocator?

nagisa commented on library/alloc/src/boxed.rs:
```diff
@@ -518,20 +751,47 @@ impl<T: ?Sized> Box<T> {
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn leak_with_alloc<'a>(b: Self) -> (&'a mut T, A)
```
This API seems sketchy to me. Without thinking too much about it I think one could do something along the lines of:

```rust
let (ref, allocref) = Box::leak_with_alloc(a_box);
drop(allocref); // potentially drops the backing storage
use(ref); // use of a dangling reference
```

TimDiekmann replied:
That's indeed a problem. Do you have a solution to this or should I remove this method? Therefore also a comment should be added to `into_raw_with_alloc`, that the allocator may have to outlive the pointer.

Amanieu replied:
I agree with @nagisa, for `leak` to be sound you need to actually leak the allocator, not return it to the user. I'd just remove that method.

`into_raw_with_alloc` is fine since it returns a raw pointer and it's not our problem if the user misuses it.

TimDiekmann commented on library/alloc/src/boxed.rs:
```diff
@@ -518,9 +756,9 @@ impl<T: ?Sized> Box<T> {
+    pub fn leak<'a>(b: Self) -> &'a mut T
     where
-        T: 'a, // Technically not needed, but kept to be explicit.
+        A: 'a,
```
I added a test, which ensures, that the allocator must not be moved while the leaked data is in use.

jackh726 commented:
Just wanted to check in to say we did talk about this today at our @rust-lang/wg-traits meeting. We worked out a design that we're hoping to get into Chalk by next release on Saturday. Then it'll just be a Chalk upgrade and this should be unblocked from that.

TimDiekmann commented:
That's awesome, thank you!

TimDiekmann commented:
Rebased onto master to include #77289

bors commented:
:umbrella: The latest upstream changes (presumably #77436) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

exrook commented on library/alloc/src/boxed.rs:
````diff
@@ -412,7 +551,59 @@ impl<T: ?Sized> Box<T> {
+    /// ```
+    /// Manually create a `Box` from scratch by using the system allocator:
+    /// ```
+    /// use std::alloc::{AllocRef, System};
````
Make tests compile
```suggestion
    /// #![feature(allocator_api)]
    /// use std::alloc::{AllocRef, Layout, System};
```

TimDiekmann replied:
Thank you, I noticed a few other doc-tests which needs adjustments. Will update then when chalk was updated.

bors commented:
:umbrella: The latest upstream changes (presumably #77466) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

TimDiekmann committed:
Support custom allocators in `Box`
Remove `Box::leak_with_alloc`


Add leak-test for box with allocator


Rename `AllocErr` to `AllocError` in leak-test


Add `Box::alloc` and adjust examples to use the new API

TimDiekmann commented:
Rebased onto master to include #77515.

@rustbot modify labels: +S-waiting-on-review -S-blocked

TimDiekmann committed:
Fix mir-opt output for 32 bit

Amanieu commented:
@bors try

bors commented:
:hourglass: Trying commit d5720bba8f5b278616a2fbd0da39478879a3f68b with merge 4a41264f4c1770bc10fe5150b615051b50292150...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d5720bba8f5b278616a2fbd0da39478879a3f68b","merge_sha":"4a41264f4c1770bc10fe5150b615051b50292150"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1218580798), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37081)
Build commit: 4a41264f4c1770bc10fe5150b615051b50292150 (`4a41264f4c1770bc10fe5150b615051b50292150`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37081","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1218580798"},"merge_sha":"4a41264f4c1770bc10fe5150b615051b50292150"} -->

Amanieu commented:
@craterbot check

TimDiekmann commented:
As far as I can see, the crater run only has one error, which is related to this PR: The [`window`-crate](https://crates.io/crates/window/0.4.0) failed to compile, as it could not infer A. This is the code snippet, which failed:
```rust
struct Wayland {
   // ...
   draw: Option<NonNull<dyn crate::Draw>>,
}

impl Wayland {
    fn connect(&mut self, draw: &mut Box<dyn crate::Draw>) {
        self.draw = NonNull::new(Box::into_raw(unsafe {
            std::mem::transmute_copy(draw)
        }));
        // ...
    }
}
```

Does this means, that any code, which constructs `Box` with `mem::transmute(_copy)` could fail? Is this even guaranteed to be backwards compatible?

In either case, the snippet should be changed to `self.draw = NonNull::new(draw)`. cc @libcala

TimDiekmann committed:
Merge branch 'master' into box-alloc

AldaronLau commented:
@TimDiekmann in the context of the window crate, just making that change doesn't actually compile.

> error[E0277]: the trait bound `std::boxed::Box<(dyn Draw + 'static)>: Draw` is not satisfied
>     --> src/wayland/mod.rs:1399:34
>      |
> 1399 |         self.draw = NonNull::new(draw);
>      |                                  ^^^^ the trait `Draw` is not implemented for `std::boxed::Box<(dyn Draw + 'static)>`
>      |
>      = note: required for the cast to the object type `dyn Draw`
> 
> error: aborting due to previous error

Which is why I wrote it that way.  play.rust-lang isn't getting the full context of the situation for some reason (because it's a different trait, maybe?).

AldaronLau commented:
@robojumper Thanks!  It's fixed now on GitHub.

TimDiekmann commented:
@Amanieu What do you think how should we proceed? Keep `into_raw` in `Box<T, A>` or move it back to `Box<T, Global>`? If we'd go with the latter, `leak` should also be moved back.

Amanieu commented:
I would prefer keeping `into_raw` in `Box<T, A>`. Since there is only one regression and it is a) related to type inference and b) already fixed, I think we can go ahead with this change.

Amanieu commented on library/alloc/src/boxed.rs:
```diff
@@ -472,13 +720,30 @@ impl<T: ?Sized> Box<T> {
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn alloc(b: &Self) -> &A {
```
This function seems poorly named to me since `alloc` sounds like a verb (`Box::alloc` sounds like it should allocate memory). Maybe rename it to `alloc_ref` or `allocator` to make it clearer.

TimDiekmann committed:
Rename `Box::alloc` to `Box::alloc_ref`

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 693a2bf18b7090202784f561de3dfca45c4f79be has been approved by `Amanieu`

<!-- @bors r=Amanieu 693a2bf18b7090202784f561de3dfca45c4f79be -->
<!-- homu: {"type":"Approved","sha":"693a2bf18b7090202784f561de3dfca45c4f79be","approver":"Amanieu"} -->

Aaron1011 commented on library/alloc/src/boxed.rs:
```diff
@@ -472,13 +720,30 @@ impl<T: ?Sized> Box<T> {
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn alloc_ref(b: &Self) -> &A {
```
To me, this sounds like we're allocating a reference, not obtaining a reference to the allocator.

memoryruins replied:
`allocator` was also suggested in https://github.com/rust-lang/rust/pull/77187#discussion_r509881386, which might be clearest.

TimDiekmann replied:
I'm fine with both, I just picked `alloc_ref` because of the trait name. Maybe we should discuss (in the WG), if we should rename `AllocRef` to `Allocator` and `alloc` to `allocate`.

TimDiekmann replied:
Proposal opened: https://github.com/rust-lang/wg-allocators/issues/76

nagisa commented on library/alloc/src/boxed.rs:
````diff
@@ -235,14 +233,103 @@ impl<T> Box<T> {
+    /// assert_eq!(*five, 5)
+    /// ```
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    // #[unstable(feature = "new_uninit", issue = "63291")]
````
Commented code should be removed before landing.

Amanieu replied:
I think the intent here is that the method requires both features to be stabilized.

nagisa replied:
Hm, I wonder if that should really be documented as a feature request to enable cfg-ish syntax to allow saying something like `unstable(all(...))` or `unstable(any(...))`.

bors commented:
:umbrella: The latest upstream changes (presumably #78350) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

TimDiekmann committed:
Merge remote-tracking branch 'upstream/master' into box-alloc

TimDiekmann commented:
Applied #78208

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

TimDiekmann commented:
Maybe rollup=never should be applied to keep track of small performance changes.

Amanieu commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 06e4497a04615ad95dff4240ca9980f19ed364ad has been approved by `Amanieu`

<!-- @bors r=Amanieu 06e4497a04615ad95dff4240ca9980f19ed364ad -->
<!-- homu: {"type":"Approved","sha":"06e4497a04615ad95dff4240ca9980f19ed364ad","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 06e4497a04615ad95dff4240ca9980f19ed364ad with merge fd542592f08ca0d1f7255600115c2eafdf6b5da7...
<!-- homu: {"type":"BuildStarted","head_sha":"06e4497a04615ad95dff4240ca9980f19ed364ad","merge_sha":"fd542592f08ca0d1f7255600115c2eafdf6b5da7"} -->

rust-log-analyzer commented:
Your PR [failed](https://github.com/rust-lang-ci/rust/runs/1311345729) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/fd542592f08ca0d1f7255600115c2eafdf6b5da7/checks/1311345729/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/fd542592f08ca0d1f7255600115c2eafdf6b5da7/checks/1311345729/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1311811632)
Approved by: Amanieu
Pushing fd542592f08ca0d1f7255600115c2eafdf6b5da7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1311811632"},"merge_sha":"fd542592f08ca0d1f7255600115c2eafdf6b5da7"} -->

TimDiekmann commented:
@rustbot ping infra

The merged succeeded, but a CI job failed. I don't know, what the job is about, but the fail is unrelated to this PR.

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#77187!

Tested on commit rust-lang/rust@fd542592f08ca0d1f7255600115c2eafdf6b5da7.
Direct link to PR: <https://github.com/rust-lang/rust/pull/77187>

💔 miri on windows: test-pass → test-fail (cc @oli-obk @eddyb @RalfJung).
💔 miri on linux: test-pass → test-fail (cc @oli-obk @eddyb @RalfJung).

RalfJung commented on library/alloc/src/boxed.rs:
```diff
@@ -472,13 +720,30 @@ impl<T: ?Sized> Box<T> {
+
+        // The box is unitiliazed later when moving out the allocator. The pointer is stored
+        // beforehand.
+        let ptr = b.0;
```
I'm afraid this code broke Stacked Borrows. The code before was very careful to go through a mutable reference, as the comment explains. That comment was retained but the mutable reference is no longer used, so this code no longer works.

I'll submit a PR to fix this. In the future, when you find comments like this and are not sure what to do about them, feel free to ping me. :)

H2CO3 commented:
I would also like to raise awareness about the issue of this breaking ABI here, [more in my comment over at the related PR for `Vec`](https://github.com/rust-lang/rust/pull/78461#issuecomment-735270859).

yoshuawuyts commented:
Is there a write-up available on this design? I've been reading through the custom allocator PRs, and don't have a good sense why for example https://github.com/rust-lang/rust/pull/58457 was abandoned in favor of the current design. As far as I can tell the current design was introduced without an RFC?

RalfJung commented:
FWIW this PR caused a lot of pain in the compiler, a long stream of ICEs, accidental UB -- altogether causing a lot of work for a lot of people. See https://github.com/rust-lang/rust/issues/95453 for more details.

This should never have landed without a discussion with the compiler team on how to support `Box` with an extra field. `Box` is a language-primitive type, its definition cannot change without adjusting many parts of the compiler and even the language spec.
